core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -305,21 +305,19 @@ private static RexNode stripCastFromString(RexNode node) {
       final RexNode o1 = call.operands.get(1);
       if (o0.getKind() == SqlKind.CAST
           && o1.getKind() != SqlKind.CAST) {
-        final RexNode o0b = ((RexCall) o0).getOperands().get(0);
-        switch (o0b.getType().getSqlTypeName()) {
-        case CHAR:
-        case VARCHAR:
-          return call.clone(call.getType(), ImmutableList.of(o0b, o1));
+        if (dialect.castRequiredForStringOperand((RexCall) o0)) {
+          return node;
```

Comment:
I'm considering if we should introduce a method like `SqlDialect#supportImplicitCast`, the `castRequiredForStringOperand` seems a little too related to the sql contexts. There are kinds of sql contexts for implicit type coercion, also not only for STRING type.

Reply:
working on SqlDialect#supportImplicitCast method